gulpfile.js/changelog.js

```diff
@@ -0,0 +1,380 @@
+				if (key) {
+					current = current[key] = current[key] || {};
+				} else {
+					(current[key] = current[key] || []).push(info);
+				}
+			}
+		}
+
+
+		/** @param {CommitChange} change */
+		function notGenerated(change) {
+			return !change.file.endsWith('.min.js') && ['prism.js', 'components.js', 'package-lock.json'].indexOf(change.file) === -1;
+		}
+		/** @param {CommitChange} change */
+		function notPartlyGenerated(change) {
+			return change.file !== 'plugins/autoloader/prism-autoloader.js' &&
+				change.file !== 'plugins/show-language/prism-show-language.js';
+		}
+		/** @param {CommitChange} change */
+		function notTests(change) {
+			return !/^tests\//.test(change.file);
+		}
+		/** @param {CommitChange} change */
+		function notExamples(change) {
+			return !/^examples\//.test(change.file);
+		}
+		/** @param {CommitChange} change */
+		function notFailures(change) {
+			return !/^known-failures.html$/.test(change.file);
+		}
+		/** @param {CommitChange} change */
+		function notComponentsJSON(change) {
+			return change.file !== 'components.json';
+		}
+
+		/**
+		 * @param {((e: T, index: number) => boolean)[]} filters
+		 * @returns {(e: T, index: number) => boolean}
+		 * @template T
+		 */
+		function and(...filters) {
+			return (e, index) => {
+				let res = true;
+				for (let i = 0, l = filters.length; i < l && res; i++) {
+					res = res && filters[i](e, index);
+				}
+				return res;
+			};
+		}
+
+		/**
+		 * Some commit message have the format `component changed: actual message`.
+		 * This function can be used to remove this prefix.
+		 *
+		 * @param {string} prefix
+		 * @param {CommitInfo} info
+		 * @returns {{ message: string, hash: string }}
+		 */
+		function removeMessagePrefix(prefix, info) {
+			const source = String.raw`^${prefix.replace(/([^-\w\s])/g, '\\$1').replace(/[-\s]/g, '[-\\s]')}:\s*`;
+			const patter = RegExp(source, 'i');
+			return {
+				message: info.message.replace(patter, ''),
+				hash: info.hash
+			};
+		}
+
+
+		/**
+		 * @type {((info: CommitInfo) => boolean)[]}
+		 */
+		const commitSorters = [
+
+			function rebuild(info) {
+				if (info.changes.length > 0 && info.changes.filter(notGenerated).length === 0) {
+					console.log('Rebuild found: ' + info.message);
+					return true;
+				}
+			},
+
+			function addedComponent(info) {
+				let relevantChanges = info.changes.filter(and(notGenerated, notTests, notExamples, notFailures));
+
+				// `components.json` has to be modified
+				if (relevantChanges.some(c => c.file === 'components.json')) {
+					relevantChanges = relevantChanges.filter(and(notComponentsJSON, notPartlyGenerated));
+
+					// now, only the newly added JS should be left
+					if (relevantChanges.length === 1) {
+						const change = relevantChanges[0];
+						if (change.mode === 'A' && change.file.startsWith('components/prism-')) {
+							var lang = change.file.match(/prism-([\w-]+)\.js$/)[1];
+							var titles = [languages[lang].title];
```

Comment:
If we're gonna es6 the gulpfile, use `let` or `const` all the way down.

Reply:
Force of habit.